### ISSUE TITLE
MicroPython REPL

### ISSUE THREAD

**Issue by iamrecursion** [ClaraCrazy/Flipper-Xtreme#572]
### Description of the feature you're suggesting.

It's all well and good to have a vast array of apps—which you guys ensure we do—but sometimes I find myself wanting to prototype or hack-together functionality directly on device.

It would be _awesome_ to have a [MicroPython](https://github.com/micropython/micropython) REPL that allows you to issue commands on-device, and provides access to the GPIO pins, radios, and so on.

**Comment by dropalltables** [comment]
that would be hard, not enough RAM on the device to store a REPL AND the OS in memory.

**Comment by iamrecursion** [comment]
You'd have to hand control directly over to micropython for the time it was running, rather than trying to run it as a flipper app.

**Comment by dropalltables** [comment]
Then how would it exit the REPL

**Comment by iamrecursion** [comment]
Hand it back to the OS by booting. Fundamentally it's just an alternate 'kernel' running on the microcontroller. May well be out of scope for this project but I figured I'd bring it up.

**Comment by dropalltables** [comment]
Hmm, interesting, do you think you could program it yourself?

**Comment by iamrecursion** [comment]
Given enough time, I think so. Unfortunately I'm very time constrained at the moment so I'm unlikely to pick it up any time soon!

**Comment by ClaraCrazy** [comment]
Funnily enough, i had that same idea a while ago, unfortunately i'm far too stupid to take on such a task myself, and the very small team I do have is very busy with other features and their private lives atm

**Comment by 1Riki** [comment]
might be a wild idea tbh but I was wondering about speaking to some Hardware oriented guys to figure out if the board and overall architecture would allow for an external attachment that can be used as a general purpose ‘outboard’ memory device which could serve as either RAM or VRAM if memory needs are exceeded on-device. Perhaps it’s possible to actually run everything that’s  in RAM including the OS in parallel and have the firmware wire-in that ‘outboard’ memory if it needs to offload some data temporarily. Correct me if I’m wrong but I think it’s possible (although I’m definitely not the best with the Hardware side of things) and I think having a parallel type arrangement would (in my mind at least) make it much faster and therefore possible move data back and forth faster while the OS on the ‘outboard’ memory device would be able to manage the handing-back of data as needed. I could be wrong of course but what do you guys reckon? A real Hardware pro would be needed; if anyone here is one then you can probably tell me if that’s possible given the board we’re working with…On 25 Mar 2024, at 02:42, JSdev ***@***.***> wrote:﻿
that would be hard, not enough RAM on the device to store a REPL AND the OS in memory.

**Comment by iamrecursion** [comment]
Cheers for the heads up @ClaraCrazy. I'll keep an eye on this one if I ever have time to poke at it.

**Comment by dropalltables** [comment]
woo! i would love it if this made it into xtreme
